File: src/main/java/org/n52/spare/kicker/resources/PlayersController.java
package org.n52.spare.kicker.resources;

import java.security.Principal;
import java.util.List;
import java.util.Optional;

import org.n52.spare.kicker.model.Player;
import org.n52.spare.kicker.model.Views;
import org.n52.spare.kicker.repositories.PlayerRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.fasterxml.jackson.annotation.JsonView;

@RestController
@RequestMapping("/players")
public class PlayersController {

	@Autowired
	private PlayerRepository playerRepository;

	@JsonView(Views.Basic.class)
	@RequestMapping("")
	public List<Player> collection(@RequestParam(required = false) Optional<Integer> page,
			@RequestParam(required = false) Optional<Integer> size) {
		return playerRepository.findAll(PageRequest.of(page.orElse(0), size.orElse(10))).getContent();
	}

	@JsonView(Views.Details.class)
	@RequestMapping("/{id}")
	public Player single(@PathVariable Long id) {
		return playerRepository.findById(id).get();
	}

	@RequestMapping("/me")
	public Player user(Principal user) {
		if (user != null) {
			Optional<Player> opt = playerRepository.findByName(user.getName());
			if (opt.isPresent()) {
				return opt.get();
			}
		}
		
		throw new IllegalStateException("Invalid user");
	}
}

File: src/main/kotlin/org/n52/spare/kicker/resources/PlayersController.kt
package org.n52.spare.kicker.resources

import java.security.Principal

import org.n52.spare.kicker.repositories.PlayerRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.PageRequest

import com.fasterxml.jackson.annotation.JsonView
import org.n52.spare.kicker.model.Match
import org.n52.spare.kicker.model.PageableResponse
import org.n52.spare.kicker.model.Player
import org.n52.spare.kicker.model.Views
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.web.bind.annotation.*
import java.util.*

@RestController
@RequestMapping("/players")
class PlayersController {

    @Autowired
    private val playerRepository: PlayerRepository? = null


    @JsonView(Views.Basic::class)
    @RequestMapping("")
    fun collection(@RequestParam(required = false) page: Optional<Int>,
                   @RequestParam(required = false) size: Optional<Int>): PageableResponse<Player> {
        val result = playerRepository!!.findAll(PageRequest.of(page.orElse(0), size.orElse(10)))
        return PageableResponse.from(result);
    }

    @JsonView(Views.Details::class)
    @RequestMapping("/{id}")
    fun single(@PathVariable id: Long?): Player {
        return playerRepository!!.findById(id!!).get()
    }

    @RequestMapping(value = "", method = arrayOf(RequestMethod.POST))
    fun createPlayer(@RequestBody player: Player?): Player {
        if (player != null) {
            if (player.id != null) {
                throw IllegalArgumentException("'id' must not be provided")
            }

            if (player.nickName == null) {
                throw IllegalArgumentException("'nickName' must be provided")
            }

            // TODO players are created on the fly atm; TODO auth layer
            player.password = BCryptPasswordEncoder().encode("kicker")

            return playerRepository!!.save<Player?>(player)
        }
        throw IllegalArgumentException("no player object provided")
    }

    @RequestMapping("/me")
    fun user(user: Principal?): Player {
        if (user != null) {
            val opt = playerRepository!!.findByName(user.name)
            if (opt.isPresent) {
                return opt.get()
            }
        }

        throw IllegalStateException("Invalid user")
    }
}
